iroh-net/src/magic_endpoint.rs

```diff
@@ -412,6 +427,39 @@ impl MagicEndpoint {
         self.msock.tracked_endpoint(node_id).await
     }
 
+    async fn resolve(&self, peer_id: &PublicKey) -> Result<AddrInfo> {
+        if let Some(discovery) = self.msock.discovery() {
+            discovery.resolve(peer_id).await
+        } else {
+            anyhow::bail!("no discovery mechanism configured");
+        }
+    }
+
+    /// Connect to a remote endpoint, using just the nodes's [`PublicKey`].
+    pub async fn connect_by_node_id(
+        &self,
+        node_id: &PublicKey,
+        alpn: &[u8],
+    ) -> Result<quinn::Connection> {
+        let addr = match self.msock.get_mapping_addr(node_id).await {
+            Some(addr) => addr,
+            None => {
+                let info = self.resolve(node_id).await?;
+                let peer_addr = NodeAddr {
+                    node_id: *node_id,
+                    info,
+                };
+                self.add_node_addr(peer_addr)?;
+                self.msock.get_mapping_addr(node_id).await.ok_or_else(|| {
+                    anyhow!("Failed to retrieve the mapped address from the magic socket. Unable to dial node {node_id:?}")
+                })?
+            }
+        };
+
+        debug!("connecting to {}: (via {})", node_id, addr);
```

Comment:
would be good to log how the addr resolution was done (for later debugging sessions..)

Reply:
You mean just "was already in peer map" / "was using the discovery mechanism" or more fine grained? The former can be done as it is, for the latter you would have to change the discovery trait.

But - proper attribution of where you got discovery info would be a larger topic. E.g. what if you get the same info from multiple sources, like mdns and pkarr? Just pick one at random? Do you keep attribution in the peer map as well?

Reply:
I added the simple option - every time discovery resolve it is used it will be logged at debug level - the discovery trait requires Debug.